### PR TITLE
Support resource files with spaces

### LINKED ISSUE #2369: Collada fails to find texture with space in file name
**Original report ([archived issue](https://osrf-migration.github.io/gazebo-gh-pages/#!/osrf/gazebo/issues/2369)) by Shane Loretz (Bitbucket: [Shane Loretz](https://bitbucket.org/%7B656e3311-aad9-45a1-aaf7-b0ee0e84b287%7D/), GitHub: [sloretz](https://github.com/sloretz)).**

The original report had attachments: [space_in_texture_filename.tar.gz](https://osrf-migration.github.io/gazebo-gh-pages/data/repositories/osrf/gazebo/issues/2369/attachments/space_in_texture_filename.tar.gz)

----------------------------------------
In gazebo 7 on ubuntu xenial a model with a collada mesh will not show a texture if it has a space in the filename. (Model attached)

The following lines appear running gazebo verbose.
```
[Err] [Material.cc:100] Unable to find texture[Material%20Ambient.png] in path[/other/dir/textured_mesh/meshes]
[Err] [SystemPaths.cc:410] File or path does not exist["/other/dir/textured_mesh/meshes/../materials/textures/Material%20Ambient.png"]
[Err] [RenderEngine.cc:483] URI doesn't exist[/other/dir/textured_mesh/meshes/../materials/textures/Material%20Ambient.png]
[Err] [Material.cc:100] Unable to find texture[Material%20Ambient.png] in path[/other/dir/textured_mesh/meshes]
```

It looks like the loader did not convert the `%20` to a space. If I change it to an underscore the texture is correctly applied. The filename of a texture in a collada file is specified using the tag [<init_from>](https://www.khronos.org/files/collada_spec_1_4.pdf) which has the type [xs:anyURI](http://www.datypic.com/sc/xsd/t-xsd_anyURI.html) which requires characters like spaces escaped with a `%`. The code responsible for fetching a material will need to un-escape the filename before searching for it.

### PR DESCRIPTION
**Pull request by chapulina**
Addresses #2870 .

Most file types were already working with whitespaces without issues. I added tests anyway to make sure they keep working. I've also only tested locally on Ubuntu, so I'm curious to see whether the new tests pass for all platforms.

Changes:

* Support encoded file names within COLLADA files (closes #2369).
* Fixed `ServerFixture` to support world names with spaces by artificially injecting `%20` into it internally.

I also took the opportunity to add tests for entity names with spaces wherever I could. There are known issues with spaces in entity names that I didn't address though, such as creating transport topics or URIs (#2056).

The only file types that are not getting support for spaces are MTL files. See this comment: https://github.com/osrf/gazebo/issues/2870#issuecomment-725058789. Technically we could update our fork of `tinyobjloader`, but it's not a good idea to diverge from upstream.

---

### Manual testing

From a clone of this branch, try the following.

Load a world with spaces:

```
gazebo --verbose "test/worlds/world with spaces.world"
```

or 

```
gazebo --verbose test/worlds/world\ with\ spaces.world
```

Load a model with spaces in various resources:

```
GAZEBO_MODEL_PATH=`pwd`/test/models/testdb gazebo --verbose
```

Then insert `Model with spaces` from the Insert menu.

### CODE REVIEW

File: test/models/testdb/model with spaces/model.sdf [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+      <link name="nested link with spaces">
+        <pose>0.25 0 0.0 0 0 0</pose>
+        <collision name="nested collision with spaces">
+          <geometry>
+            <box>
+              <size>1 1 1</size>
+            </box>
+          </geometry>
+        </collision>
+        <visual name="neste visual with spaces">
```
Comment (iche033) [inline test/models/testdb/model with spaces/model.sdf]: `neste` -> `nested`

**Review by iche033 — COMMENTED** [review]
looks good. Also tested adding a sensor and light with space in the name and the model loads fine. 

A few issues I noticed that can be addressed later if needed:

* I see a few error msgs from ign-transport - looks like it does not allow spaces in the topic name:

```
Topic [/included model/joint_cmd] is not valid.
[Err] [JointController.cc:66] Error subscribing to topic [/included model/joint_cmd]
```

* the plotting tool may not be working with spaces yet. It prints out a bunch of URI warnings, e.g.:

```
[Wrn] [URI.cc:327] Unable to parse URI [data://world/world with spaces/model/included model/joint/joint between links nested at different depths?p=axis/0/double/position]. Ignoring.
```

* A couple of factory test failures that look related on [homebrew](https://build.osrfoundation.org/job/gazebo-ci-pr_any-homebrew-amd64/2206/):

```
FactoryTest.ActorSpaces
FactoryTest.FilenameFuelURL
```

**Comment by chapulina** [comment]
> A few issues I noticed that can be addressed later if needed:

Yeah that's https://github.com/osrf/gazebo/issues/2056. As I mentioned on the description above, this is out of scope for this PR. The focus here are spaces on the filenames. I took the opportunity to also test spaces in entity names wherever it was already working, but I'm not planning to fix the remaining issues in this PR.

> FactoryTest.ActorSpaces

I also noticed that was flaky for me locally. The changes in f8fe4a3 seem to have fixed it.

> FactoryTest.FilenameFuelURL

That looks flaky from before: https://build.osrfoundation.org/job/gazebo-ci-pr_any-homebrew-amd64/2206/testReport/(root)/FactoryTest/FilenameFuelURL/history/

File: test/models/testdb/model with spaces/model.sdf (hunk 1 above)
Comment (chapulina) [reply test/models/testdb/model with spaces/model.sdf]: aha, that's what happens when tests aren't thorough. Fixed and added more tests in f8fe4a3

File: test/integration/factory.cc [hunk 2]
```diff
@@ -1146,6 +1214,199 @@ TEST_F(FactoryTest, FilenameModelDatabase)
...
+      visualVis->GetMeshName());
+
+  // Nested
+  EXPECT_EQ(1u, model->NestedModels().size());
+  auto nestedModel = model->NestedModel("nested model with spaces");
+  ASSERT_NE(nullptr, nestedModel);
+
+  auto nestedModelVis = modelVis->GetChild(2);
+  ASSERT_NE(nullptr, nestedModelVis);
+  EXPECT_EQ("model with spaces::nested model with spaces", nestedModelVis->Name());
```
Comment (iche033) [inline test/integration/factory.cc]: minor style issue: line over 80 char

**Comment by iche033** [comment]
> FactoryTest.FilenameFuelURL

ah it's the SSL certificate issue. This fixes the test for me:

```
@@ -1419,7 +1420,7 @@ TEST_F(FactoryTest, FilenameFuelURL)
 
   msgs::Factory msg;
   msg.set_sdf_filename(
-      "https://api.ignitionfuel.org/1.0/chapulina/models/Test box");
+      "https://fuel.ignitionrobotics.org/1.0/chapulina/models/Test box/1");
```

I had to include the version number in the URL otherwise the gazebo complains about not being able to find file in `Test box/tip` directory

**Review by iche033 — APPROVED** [review]
looks good, just one minor style issue. We can address the `FactoryTest.FilenameFuelURL` issue here or separately. Either way is fine.

**Pull request merged**